### PR TITLE
YARA Integration

### PR DESCRIPTION
**Pull request by theopolis**
This is a squash of @wxsBSD's `yara2` branch.

### CODE REVIEW

**Comment by theopolis** [comment]
Hey @wxsBSD, I'm getting build errors on OS X:

```
> ~/git/github/osquery(branch:yara2) » SKIP_TESTS=1 make VERBOSE=1
export CTEST_OUTPUT_ON_FAILURE=1
mkdir -p build/darwin
cd build/darwin && cmake ../.. && \
		CTEST_OUTPUT_ON_FAILURE=1 make --no-print-directory VERBOSE=1
Re-run cmake no build system arguments
-- Building for OS X
-- Found components for DL
-- Found readline library
-- Found RocksDB
-- Thrift version 0.9.2
-- Found library dependency /usr/local/lib/libboost_thread-mt.a
-- Found library dependency /usr/local/lib/liblz4.a
-- Found library dependency /usr/local/lib/libboost_system.a
-- Found library dependency /usr/local/lib/libboost_filesystem.a
-- Found library dependency /usr/local/lib/libboost_regex.a
-- Configuring done
-- Generating done
-- Build files have been written to: /Users/reed/git/github/osquery/build/darwin
/usr/local/Cellar/cmake/3.1.0/bin/cmake -H/Users/reed/git/github/osquery -B/Users/reed/git/github/osquery/build/darwin --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/Cellar/cmake/3.1.0/bin/cmake -E cmake_progress_start /Users/reed/git/github/osquery/build/darwin/CMakeFiles /Users/reed/git/github/osquery/build/darwin/CMakeFiles/progress.marks
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/Makefile2 all
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/libglog.dir/build.make CMakeFiles/libglog.dir/depend
cd /Users/reed/git/github/osquery/build/darwin && /usr/local/Cellar/cmake/3.1.0/bin/cmake -E cmake_depends "Unix Makefiles" /Users/reed/git/github/osquery /Users/reed/git/github/osquery /Users/reed/git/github/osquery/build/darwin /Users/reed/git/github/osquery/build/darwin /Users/reed/git/github/osquery/build/darwin/CMakeFiles/libglog.dir/DependInfo.cmake --color=
Dependee "/Users/reed/git/github/osquery/build/darwin/CMakeFiles/libglog.dir/DependInfo.cmake" is newer than depender "/Users/reed/git/github/osquery/build/darwin/CMakeFiles/libglog.dir/depend.internal".
Scanning dependencies of target libglog
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/libglog.dir/build.make CMakeFiles/libglog.dir/build
make[3]: Nothing to be done for `CMakeFiles/libglog.dir/build'.
/usr/local/Cellar/cmake/3.1.0/bin/cmake -E cmake_progress_report /Users/reed/git/github/osquery/build/darwin/CMakeFiles  4 5 6 7
[  4%] Built target libglog
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/yara.dir/build.make CMakeFiles/yara.dir/depend
cd /Users/reed/git/github/osquery/build/darwin && /usr/local/Cellar/cmake/3.1.0/bin/cmake -E cmake_depends "Unix Makefiles" /Users/reed/git/github/osquery /Users/reed/git/github/osquery /Users/reed/git/github/osquery/build/darwin /Users/reed/git/github/osquery/build/darwin /Users/reed/git/github/osquery/build/darwin/CMakeFiles/yara.dir/DependInfo.cmake --color=
Dependee "/Users/reed/git/github/osquery/build/darwin/CMakeFiles/yara.dir/DependInfo.cmake" is newer than depender "/Users/reed/git/github/osquery/build/darwin/CMakeFiles/yara.dir/depend.internal".
Scanning dependencies of target yara
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/yara.dir/build.make CMakeFiles/yara.dir/build
/usr/local/Cellar/cmake/3.1.0/bin/cmake -E cmake_progress_report /Users/reed/git/github/osquery/build/darwin/CMakeFiles 100
[  5%] Performing bootstrap step for 'yara'
cd /Users/reed/git/github/osquery/third-party/yara-3.3.0 && sh ./bootstrap.sh
glibtoolize: putting auxiliary files in `.'.
glibtoolize: copying file `./ltmain.sh'
glibtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
glibtoolize: copying file `m4/libtool.m4'
glibtoolize: copying file `m4/ltoptions.m4'
glibtoolize: copying file `m4/ltsugar.m4'
glibtoolize: copying file `m4/ltversion.m4'
glibtoolize: copying file `m4/lt~obsolete.m4'
configure.ac:28: error: possibly undefined macro: AC_MSG_ERROR
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
autoreconf: /usr/local/Cellar/autoconf/2.69/bin/autoconf failed with exit status: 1
make[3]: *** [yara-prefix/src/yara-stamp/yara-bootstrap] Error 1
make[2]: *** [CMakeFiles/yara.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [all] Error 2
```

The important, yet pretty vague line: `configure.ac:28: error: possibly undefined macro: AC_MSG_ERROR`. 

Did you run into this autoreconf/pkg-config/libtoolize issue?

**Comment by theopolis** [comment]
Context:
```
> autoreconf -V
autoreconf (GNU Autoconf) 2.69
Copyright (C) 2012 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<http://gnu.org/licenses/gpl.html>, <http://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.

> brew info pkg-config
pkg-config: stable 0.28 (bottled)
http://pkgconfig.freedesktop.org
/usr/local/Cellar/pkg-config/0.28 (10 files, 604K) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/pkg-config.rb

> brew info autoconf
autoconf: stable 2.69 (bottled)
https://www.gnu.org/software/autoconf
/usr/local/Cellar/autoconf/2.69 (69 files, 2.0M) *
  Poured from bottle
From: https://github.com/Homebrew/homebrew/blob/master/Library/Formula/autoconf.rb

**Comment by wxsBSD** [comment]
I did not run into it on OS X, but I did on Ubuntu, and now that you mention this it explains the build problem I was having on Ubuntu.

Here's my ubuntu VM:
```
wxs@vm:~$ autoreconf -V
autoreconf (GNU Autoconf) 2.69
Copyright (C) 2012 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<http://gnu.org/licenses/gpl.html>, <http://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.
wxs@vm:~$ 
```

Here's my OS X box. Note that I don't use brew, mostly because I have a lot of things that break badly when using it.
```
wxs@psh wxs % autoreconf -V
autoreconf (GNU Autoconf) 2.68
Copyright (C) 2010 Free Software Foundation, Inc.
License GPLv3+/Autoconf: GNU GPL version 3 or later
<http://gnu.org/licenses/gpl.html>, <http://gnu.org/licenses/exceptions.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by David J. MacKenzie and Akim Demaille.
wxs@psh wxs % 
```

I suspect whatever the fix is for you will also be the fix for me. :)

**Comment by theopolis** [comment]
Cool, yeah we can figure out a name, "yara" wont cover it if we ever want to expose additional yara-related details, like the set of signatures/files in use. I'll figure out the build details then you can take over my branch, and re-PR? ;)

**Comment by theopolis** [comment]
Hey @wxsBSD, I just pulled from their master :/ I couldn't figure out what was erroring with autotools, though their most recent changes to autoreconf/configure were small and picking them singularly didn't fix the missing macro (which is a very esoteric error usually indicative of something wrong with libtool installs).

It doesn't seem like there was anything *that* wrong with our default builds.

**Comment by javuto** [comment]
@osqueryer, ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by wxsBSD** [comment]
```
Cool, yeah we can figure out a name, "yara" wont cover it if we ever want to expose additional yara-related details, like the set of signatures/files in use. I'll figure out the build details then you can take over my branch, and re-PR? ;)
```

That sounds good to me. I could see myself one day experimenting with the ability to scan a given PID from osquery. :)

As for taking it over, at this point I don't have anything else to add. If we can work out the build issues I'd be happy to just see it merged. If there are other things blocking it I will certainly take this back over and address them.

**Comment by theopolis** [comment]
ok to test

**Comment by javuto** [comment]
ok to test

**Comment by javuto** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

File: include/osquery/events.h [hunk 1]
```diff
@@ -20,7 +20,6 @@
 #include <boost/thread/locks.hpp>
 #include <boost/thread/mutex.hpp>
 
-#include <osquery/database.h>
```
Comment (theopolis) [inline include/osquery/events.h]: This was the only hashmap dependency I could find, which comes from RocksDB. Moving this from the public headers into the implementation removes the max/min overload from YARA until their next >3.3.0 release.

**Comment by theopolis** [comment]
Waiting for tests from @wxsBSD, there might be an issue/bug with the usage. We can land this PR then add unit tests in the next PR followed by documentation.

**Comment by wxsBSD** [comment]
I've pulled in these changes into a branch at https://github.com/wxsBSD/osquery/tree/theopolis_yara2, and then added a couple more commits (https://github.com/wxsBSD/osquery/commit/5bdb89448e6e3c5621d2642ca2cbc0fc6c70cf2a and https://github.com/wxsBSD/osquery/commit/a5f4ab648afca6e894acde0d304f05d39caf8440) which fixes some problems. It passed some basic tests on my end and anything problems we can shake out once it is merged.

I went to create a PR for that branch when I noticed it wouldn't merge cleanly to master, so I went ahead and merged master into ```theopolis_yara2```, fixed conflicts and tested the build again and it failed on OS X. The failure appears to be in some of the more recent stuff and not in anything YARA related.

```
/Users/wxs/Documents/dev/osquery/build/darwin/generated/additional_amalgamation.cpp:1418:7: error: 
      redefinition of 'xattrWhereFromTablePlugin'
class xattrWhereFromTablePlugin : public TablePlugin {
      ^
/Users/wxs/Documents/dev/osquery/build/darwin/generated/additional_amalgamation.cpp:1395:7: note: 
      previous definition is here
class xattrWhereFromTablePlugin : public TablePlugin {
      ^
/Users/wxs/Documents/dev/osquery/build/darwin/generated/additional_amalgamation.cpp:1436:1: error: 
      redefinition of 'xattrWhereFromTablePluginRegistryItem'
REGISTER(xattrWhereFromTablePlugin, "table", "xattr_where_from");
^
/Users/wxs/Documents/dev/osquery/include/osquery/registry.h:79:44: note: 
      expanded from macro 'REGISTER'
  __attribute__((constructor)) static void type##RegistryItem() { \
                                           ^
<scratch space>:32:1: note: expanded from here
xattrWhereFromTablePluginRegistryItem
^
/Users/wxs/Documents/dev/osquery/build/darwin/generated/additional_amalgamation.cpp:1413:1: note: 
      previous definition is here
REGISTER(xattrWhereFromTablePlugin, "table", "xattr_where_from");
^
/Users/wxs/Documents/dev/osquery/include/osquery/registry.h:79:44: note: 
      expanded from macro 'REGISTER'
  __attribute__((constructor)) static void type##RegistryItem() { \
                                           ^
<scratch space>:31:1: note: expanded from here
xattrWhereFromTablePluginRegistryItem
^
2 errors generated.
```

At this point feel free to either grab my changes from ```theopolis_yara2``` or let me know if you want me to open a PR for it. Thanks again for the help with the build stuff!

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Comment by theopolis** [comment]
ok to test

**Pull request merged**